group/assignment5/UnknownAlphabet.java

```diff
@@ -0,0 +1,184 @@
+// During topological traversal, chars will be saved in reversed order in a stack reversed_alphabet.
+// Used stack to have the correct order using the last-in-first-out property.
+// Keep a list vertices of all the vertices created from the words in dictionary.
+// Keep a map of every char to its vertex to check if a vertex was already created for that char.
+public class UnknownAlphabet {
+
+	Stack<Character> reversed_alphabet = new Stack<Character>();
+	LinkedList<Vertex> vertices = new LinkedList<Vertex>();
+	HashMap<Character, Vertex> char_vertex_map = new HashMap<Character, Vertex>();
+	
+	// Assume dictionary input is a List.
+	public LinkedHashSet<Character> getAlphabet(List<String> dict) throws Exception {
+
+		LinkedHashSet<Character> ordered_alphabet = new LinkedHashSet<Character>();
+
+		if (dict.isEmpty())
+			return ordered_alphabet;
+
+		// Return all characters (no duplicates) if dictionary has only one word.
+		if (dict.size() == 1) {
+			String word = dict.get(0);
+			for (int i = 0; i < word.length(); i++) {
+				ordered_alphabet.add(word.charAt(i));
+			}
+			return ordered_alphabet;
+		}
+
+		createVertices(dict);
+		topologicalSort();
+
+		// Return the characters in the correct order by reversing the stack.
+		while (!reversed_alphabet.isEmpty()) {
+			ordered_alphabet.add(reversed_alphabet.pop());
+		}
+
+		return ordered_alphabet;
+	}
+
+	// Traverse verticies which have not been visited yet.
+	private void topologicalSort() throws Exception {
+
+		for (Vertex vertex : vertices) {
+			if (vertex.getState().equals("white")) {
+				sortFromVertex(vertex);
+			}
+		}
+	}
+
+	// For each unvisisted vertex, visit its children if they haven't been visited (white).
+	// If during the visit of a child we discover it was started but not finished (grey),
+	// then we have a cycle, so no order is possible.
+	private void sortFromVertex(Vertex cur_vertex) throws Exception {
+
+		cur_vertex.setState("grey"); // starting visit, but not finished
+		for (Vertex child : cur_vertex.getChildren()) {
+			if (child.getState().equals("white")) {
+				sortFromVertex(child);
+			} else if (child.getState().equals("grey")) {
+				throw new Exception("There is a cycle, order impossible.");
+			}
+		}
+		cur_vertex.setState("black"); // finished visit.
+		reversed_alphabet.add(cur_vertex.getValue());
+	}
+
+	// Look at every 2 consecutive words to get the relation between the first 2 different characters.
+	// For example, in [ABCD, ABDE, EFGH, FMNO]: C < D (from first 2 words), A < E (from 2nd and 3rd), E < F (from 3rd and 4th)
+	// Because A < E and E < F then from transitivity => A < E without needing to check word 1 or 2 with word 4.
+	public LinkedList<Vertex> createVertices(List<String> dict) {
+
+		for (int i = 0; i < dict.size() - 1; i++) {
+			String first = dict.get(i);
+			String second = dict.get(i + 1);
+			createVerticesHelper(first, second);
+		}
+		
```

Comment:
When I was going through your solution for the first time, it was very confusing to me that you return ``vertices`` here without seemingly adding to them. I realise of course that you add the new vertex at the end of the "chain" of calls, and doing it this way is much clearer then passing ``vertices`` along all the time. So, maybe add a comment about it, or just make the method void, since you don't really need it to return a value?

Reply:
Thank you for your suggestion. I think it is cleaner to have it return void, I made it return the list for testing purposes but it might not be the best approach. Maybe a getter function to return the list of vertices for testing would be better?